src/compress/qzip.cpp

```diff
@@ -989,9 +1016,55 @@ QByteArray QZipReader::fileData(const QString &fileName) const
     d->device->seek(start);
     LocalFileHeader lh;
     d->device->read((char *)&lh, sizeof(LocalFileHeader));
-    uint skip = readUShort(lh.file_name_length) + readUShort(lh.extra_field_length);
+    uint skip = readUShort(lh.file_name_length);
     d->device->seek(d->device->pos() + skip);
 
+    // read extra fields
+    ushort extra_field_length = readUShort(lh.extra_field_length);
+    while (extra_field_length > sizeof(ExtensibleDataField)) {
+        ExtensibleDataField df;
+
+        d->device->read((char *)&df, sizeof(ExtensibleDataField));
+        extra_field_length -= sizeof(ExtensibleDataField);
+
+        ushort header = readUShort(df.header);
+        ushort data_size = readUShort(df.data_size);
+
+        if (extra_field_length < data_size) {
+            qWarning("QZip: Invalid extra field length");
+            return QByteArray();
+        }
+
+        extra_field_length -= data_size;
+
+        if (header == 0x0001) { // ZIP64
+            const QByteArray ba = d->device->read(data_size);
+            if (ba.length() != data_size) {
+                qWarning() << "QZip: Failed to read extra field in zip file, file may be incomplete";
+                return QByteArray();
+            }
+
+            const uchar *data = (const uchar *) ba.constData();
+
+            if (uncompressed_size == -1) {
+                uncompressed_size = readUInt64(data);
+                data += 8;
+            }
+            if (compressed_size == -1) {
+                compressed_size = readUInt64(data);
+                data += 8;
+            }
+            if (start == -1) {
+                start = readUInt64(data);
+                data += 8;
+            }
```

Comment:
Since you just merged this I was reading this code again, and noticed that the above three reads may go beyond the actual size of the data for invalid ZIP archives. Small issue but may be worth adding a check for.